runtime/lua/vim/filetype.lua

```diff
@@ -1170,57 +1170,57 @@ local extension = {
   -- Ignored extensions
   bak = function(path, bufnr)
     local root = vim.fn.fnamemodify(path, ':r')
-    return M.match(root, bufnr)
+    return M.match({ filename = root })
```

Comment:
Is the filename enough here?

Reply:
Oh, good call... No, I don't think so, since `root.bak` can be anything so might require looking at the contents.

I think you found a flaw in our current strategy :( We need to allow passing `filename` in addition to `buf` to `vim.match`, in which case `filename` overrides the actual name of the buffer. (Luckily this is not a breaking change at this point...)

@gpanders

Reply:
Ok, so let's lift the requirement that `buf`, `filename`, and `contents` are mutually exclusive. Instead it works like this:

When `buf` is given, `filename` and `contents` are inferred from the buffer if omitted. If `buf` is missing, one of `filename` or `contents` is required.

Supplying both `filename` and `contents` together without `buf` could be a good way to do matching without a buffer, but the problem is that our filetype detection functions get the contents from the buffer number: they don't accept just a raw contents string (or table). So currently we would have to create a new, empty buffer, write the supplied contents to that buffer, and use that new buffer in the detection functions.

Reply:
https://github.com/neovim/neovim/pull/19114

Reply:
(which is merged now, so conflicts again... sorry 😞